# Table of Content
- [Import Libraries](#import-libraries)
- [Resources](#resources)
- [Load Data Sets](#load-data-sets)
- [Data Wrangling - Orders Data](#data-wrangling---orders-data)
- [Explore Orders Data](#explore-orders-data)
- [Data Wrangling - Departments Data](#data-wrangling---departments-data)
- [Explore Products Data](#explore-products-data)
- [Export Wrangled Data Sets](#export-wrangled-data-sets)


## Import Libraries [#](#table-of-content)

In [1]:
from pathlib import Path

import numpy as np
import pandas as pd


## Resources [#](#table-of-content)

In [2]:
# project folder
project_path = Path(r"C:\Users\vynde\Desktop\CareerFoundry Data Analytics\Data Immersion - 4 Python Fundamentals for Data Analysts\Instacart_Basket_Analysis")

# resource folders
original_data_path = project_path / "02_Data" / "Original_Data"
prepared_data_path = project_path / "02_Data" / "Prepared_Data"

# input files
oders_data_path = original_data_path / "orders.csv"
departments_data_path = original_data_path / "departments.csv"

# output files
orders_wrangled_data_path = prepared_data_path / "orders_wrangled.csv"
departments_wrangled_data_path = prepared_data_path / "departments_wrangled.csv"

## Load Data Sets [#](#table-of-content)

Orders

In [3]:
# import orders data set
df_ords = pd.read_csv(original_data_path / "orders.csv")
df_ords.head()

order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0   2539329        1    prior             1          2                  8   
1   2398795        1    prior             2          3                  7   
2    473747        1    prior             3          3                 12   
3   2254736        1    prior             4          4                  7   
4    431534        1    prior             5          4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

Products

In [4]:
# import products data set
df_prods = pd.read_csv(original_data_path / "products.csv", index_col=False)
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

Departments

In [5]:
# import departments
df_dep = pd.read_csv(departments_data_path, index_col=False)
df_dep.head()

department_id       1      2       3        4        5              6  \
0    department  frozen  other  bakery  produce  alcohol  international   

           7     8                9  ...            12      13         14  \
0  beverages  pets  dry goods pasta  ...  meat seafood  pantry  breakfast   

             15          16         17      18      19    20       21  
0  canned goods  dairy eggs  household  babies  snacks  deli  missing  

[1 rows x 22 columns]

## Data Wrangling - Orders Data [#](#table-of-content)

Dropping columns

In [6]:
df_ords = df_ords.drop(columns = ['eval_set'])

Renaming columns

In [7]:
df_ords.rename(columns = {'order_dow' : 'orders_day_of_week'}, inplace = True)

Check for more unecessary variables

In [8]:
df_ords.dtypes

order_id                    int64
user_id                     int64
order_number                int64
orders_day_of_week          int64
order_hour_of_day           int64
days_since_prior_order    float64
dtype: object

Change data types if needed

In [9]:
df_ords['days_since_prior_order'] = df_ords['days_since_prior_order'].astype('Int64')  # can handle nan
df_ords.dtypes

order_id                  int64
user_id                   int64
order_number              int64
orders_day_of_week        int64
order_hour_of_day         int64
days_since_prior_order    Int64
dtype: object

## Explore Orders Data [#](#table-of-content)

Busiest hour for placing orders

In [10]:
# get counts of hours
df_ords.order_hour_of_day.value_counts().head(3)

10    288418
11    284728
15    283639
Name: order_hour_of_day, dtype: int64

>The busiest order hour is 10 am.

## Data Wrangling - Departments Data [#](#table-of-content)

Reshaping dataframe

In [11]:
# transpose and reset header of departments data frame
df_dep_t = df_dep.T
new_header = df_dep_t.iloc[0]
df_dep_t.reset_index(drop=True, inplace=True)
df_dep_t_new = df_dep_t[1:]
df_dep_t_new.columns = new_header
df_dep_t_new.head()

department_id department
1                 frozen
2                  other
3                 bakery
4                produce
5                alcohol

Data dictionary for departments in products data - What's department_id 4

In [12]:
# create data dictionary
data_dict = df_dep_t_new.to_dict('index')
data_dict[4]

{'department': 'produce'}

In [13]:
# create data dictionary V2
data_dict2 = {int(dep_id): dep for dep_id, dep in zip(df_dep.columns[1:], df_dep.values[0][1:])}
data_dict2[4]

'produce'

>department_id 4 refers to produce department

## Explore Products Data [#](#table-of-content)

Data subset containing only breakfeast items

In [14]:
# subset containing only breakfast products
df_breakfast_prods = df_prods[df_prods.department_id==14]
df_breakfast_prods

product_id                                       product_name  \
27             28                                  Wheat Chex Cereal   
33             34                                                NaN   
67             68                            Pancake Mix, Buttermilk   
89             90                                       Smorz Cereal   
210           211   Gluten Free Organic Cereal Coconut Maple Vanilla   
...           ...                                                ...   
49330       49326                            Cereal Variety Fun Pack   
49395       49391            Light and Fluffy Buttermilk Pancake Mix   
49547       49543                          Chocolate Cheerios Cereal   
49637       49633               Shake 'N Pour Buttermilk Pancake Mix   
49667       49663  Ultra Protein Power Crunch Peanut Butter N' Ho...   

       aisle_id  department_id  prices  
27          121             14    10.1  
33          121             14    12.2  
67          130             14    13.7  
89          121             14     3.9  
210         130             14     3.6  
...         ...            ...     ...  
49330       121             14     9.1  
49395       130             14     2.0  
49547       121             14    10.8  
49637       130             14    14.2  
49667        57             14    10.2  

[1116 rows x 5 columns]

Data subset containing products used for dinner parties (products from departments: alcohol, deli, beverages, and meat/seafood)

In [15]:
# subset containing alcohol, deli, beverages, and meat/seafood
df_dinner_party_prods = df_prods[df_prods.department_id.isin([5, 20, 7, 12])]
df_dinner_party_prods

product_id                                    product_name  aisle_id  \
2               3            Robust Golden Unsweetened Oolong Tea        94   
6               7                  Pure Coconut Water With Orange        98   
9              10  Sparkling Orange Juice & Prickly Pear Beverage       115   
10             11                               Peach Mango Juice        31   
16             17                               Rendered Duck Fat        35   
...           ...                                             ...       ...   
49676       49672                          Cafe Mocha K-Cup Packs        26   
49679       49675             Cinnamon Dolce Keurig Brewed K Cups        26   
49680       49676                          Ultra Red Energy Drink        64   
49686       49682                              California Limeade        98   
49688       49684       Vodka, Triple Distilled, Twist of Vanilla       124   

       department_id  prices  
2                  7     4.5  
6                  7     4.4  
9                  7     8.4  
10                 7     2.8  
16                12    17.1  
...              ...     ...  
49676              7     6.5  
49679              7    14.0  
49680              7    14.5  
49686              7     4.3  
49688              5     5.3  

[7650 rows x 5 columns]

Total counts of last subset data

> 7650

Examine strange data for customer with user_id 1

In [16]:
# subset for user 1
df_ords[df_ords.user_id=="1"]

Empty DataFrame
Columns: [order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

>order 8 and 9 are only 2 hours apart

Basic stats of user 1

In [20]:
# basic order stats for user 1
df_ords[df_ords.user_id==1].describe()

order_id  user_id  order_number  orders_day_of_week  \
count  1.100000e+01     11.0     11.000000           11.000000   
mean   1.923450e+06      1.0      6.000000            2.636364   
std    1.071950e+06      0.0      3.316625            1.286291   
min    4.315340e+05      1.0      1.000000            1.000000   
25%    8.690170e+05      1.0      3.500000            1.500000   
50%    2.295261e+06      1.0      6.000000            3.000000   
75%    2.544846e+06      1.0      8.500000            4.000000   
max    3.367565e+06      1.0     11.000000            4.000000   

       order_hour_of_day  days_since_prior_order  
count          11.000000               10.000000  
mean           10.090909               19.000000  
std             3.477198                9.030811  
min             7.000000                0.000000  
25%             7.500000               14.250000  
50%             8.000000               19.500000  
75%            13.000000               26.250000  
max            16.000000               30.000000

>The user has made 11 orders.
The orders were made from Mondays to Thursdays between 7am and 4pm.
The average days between 2 orders is 19 days which is also very close to the median of 19.5


## Export Wrangled Data Sets [#](#table-of-content)

Wrangled Orders Data

In [18]:
df_ords.to_csv(orders_wrangled_data_path)

Wrangled Departments Data

In [19]:
df_dep_t_new.to_csv(departments_wrangled_data_path)